include/seqan3/range/metafunction.hpp

```diff
@@ -0,0 +1,170 @@
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides various metafunctions used by the range module.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ * \ingroup range
+ */
+
+#pragma once
+
+#include <type_traits>
+
+#include <range/v3/utility/iterator_traits.hpp>
+#include <range/v3/range_traits.hpp>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3
+{
+
+/*!\addtogroup range
+ * \{
+ */
+
+// ----------------------------------------------------------------------------
+
+/*!\brief Whether a type has a `value_type` member and/or `ranges::value_type_t<t>` resolves [Value metafunction].
+ * \tparam t The type to recurse on; must have `ranges::value_type_t<t>`
+ * \hideinitializer
+ */
+template <typename t>
+constexpr bool has_value_type_v = false;
+
+//!\cond
+template <typename t>
+    requires requires (t) { typename ranges::value_type<std::decay_t<t>>::type; }
+constexpr bool has_value_type_v<t> = true;
+//!\endcond
+
+// ----------------------------------------------------------------------------
+
+/*!\brief Recursively determines the `value_type` on containers and/or iterators [Type metafunction].
+ * \tparam t The type to recurse on; must have `ranges::value_type_t<t>`
+ * \hideinitializer
+ */
+template <typename t>
+//!\cond
+    requires has_value_type_v<t>
+//!\endcond
+struct innermost_value_type
+{
+    //!\brief The forwarded type.
+    using type = ranges::value_type_t<std::decay_t<t>>;
+};
+
+//!\cond
+template <typename t>
+    requires has_value_type_v<t> &&
+             has_value_type_v<ranges::value_type_t<std::decay_t<t>>>
+struct innermost_value_type<t>
+{
+    using type = typename innermost_value_type<ranges::value_type_t<std::decay_t<t>>>::type;
+};
+//!\endcond
+
+// ----------------------------------------------------------------------------
+
+//!\brief Shortcut for seqan3::innermost_value_type.
+template <typename t>
+//!\cond
+    requires has_value_type_v<t>
+//!\endcond
+using innermost_value_type_t = typename innermost_value_type<t>::type;
+
+// ----------------------------------------------------------------------------
+
+/*!\brief Returns the number of times you can call `ranges::value_type_t` recursively on t [Value metafunction].
+ * \tparam t The type to be queried; must resolve `ranges::value_type_t` at least once.
+ * \hideinitializer
+ */
+template <typename t>
+//!\cond
+    requires has_value_type_v<t>
+//!\endcond
+constexpr size_t dimension_v = 1;
+
+//!\cond
+template <typename t>
+    requires has_value_type_v<t> &&
+             has_value_type_v<ranges::value_type_t<std::decay_t<t>>>
+constexpr size_t dimension_v<t> = dimension_v<ranges::value_type_t<std::decay_t<t>>> + 1;
```

Comment:
Maybe call it cardinality. I think the range library uses this term for the nesting level of ranges.
What do you think?